src/Client.php

```diff
@@ -64,54 +65,74 @@ private function createDefaultDecider(int $maxRetries): Closure
                 return false;
+            } elseif ($response && $response->getStatusCode() >= 500) {
+                $this->logger->notice(sprintf(
+                    'Got a %s response for this reason: %s, retrying.',
+                    $response->getStatusCode(),
+                    $response->getReasonPhrase()
+                ));
+                return true;
             } elseif ($error && $error->getCode() >= 500) {
+                $this->logger->notice(sprintf(
+                    'Got a %s error with this message: %s, retrying.',
+                    $error->getCode(),
+                    $error->getMessage()
+                ));
                 return true;
             } else {
                 return false;
             }
         };
     }
 
-    abstract protected function getTokenHeaderName(): ?string;
-
-    private function initClient(string $url, ?string $token, array $options = []): GuzzleClient
+    /**
+     * @param array{
+     *     handler?: HandlerStack,
+     *     backoffMaxTries: int<0, 100>,
+     *     userAgent: string,
+     *     logger?: LoggerInterface
+     * } $options
+     */
+    private function initClient(string $url, ?string $token, array $options): GuzzleClient
     {
         // Initialize handlers (start with those supplied in constructor)
         $handlerStack = HandlerStack::create($options['handler'] ?? null);
         // Set exponential backoff
         $handlerStack->push(Middleware::retry($this->createDefaultDecider($options['backoffMaxTries'])));
         // Set handler to set default headers
-        $handlerStack->push(Middleware::mapRequest(
-            function (RequestInterface $request) use ($token, $options) {
-                $request = $request
-                    ->withHeader('User-Agent', $options['userAgent'])
-                    ->withHeader('Content-type', 'application/json');
-                if ($this->getTokenHeaderName()) {
-                    $request = $request->withHeader((string) $this->getTokenHeaderName(), (string) $token);
-                }
-                return $request;
-            }
-        ));
+        $headers = $this->defaultHeaders;
+        $headers['User-Agent'] = $options['userAgent'];
+        if ($this->tokenHeaderName) {
+            $headers[$this->tokenHeaderName] = (string) $token;
+        }
+
         // Set client logger
-        if (isset($options['logger']) && $options['logger'] instanceof LoggerInterface) {
+        if (isset($options['logger'])) {
             $handlerStack->push(Middleware::log(
                 $options['logger'],
                 new MessageFormatter(
                     '{hostname} {req_header_User-Agent} - [{ts}] "{method} {resource} {protocol}/{version}"' .
                     ' {code} {res_header_Content-Length}'
-                )
+                ),
+                'debug'
             ));
+            $this->logger = $options['logger'];
```

Comment:
`$this->logger` je tu jestli dobre vidim jen kvuli logovani pri retry? Mozna neni potreba z toho delat class property, muzes si to predat jako argument do volani `createDefaultDecider` o par radek vys.